src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -0,0 +1,3129 @@
+                if (!needsCulture)
+                {
+                    int[] codes = rm.Code.Codes;
+                    for (int codepos = 0; codepos < codes.Length; codepos += RegexCode.OpcodeSize(codes[codepos]))
+                    {
+                        if ((codes[codepos] & RegexCode.Ci) == RegexCode.Ci)
+                        {
+                            needsCulture = true;
+                            break;
+                        }
+                    }
+                }
+
+                if (needsCulture)
+                {
+                    writer.WriteLine("TextInfo textInfo = CultureInfo.CurrentCulture.TextInfo;");
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static bool UseToLowerInvariant(bool hasTextInfo, RegexOptions options) => !hasTextInfo || (options & RegexOptions.CultureInvariant) != 0;
+
+        private static string ToLower(bool hasTextInfo, RegexOptions options, string expression) => UseToLowerInvariant(hasTextInfo, options) ? $"char.ToLowerInvariant({expression})" : $"textInfo.ToLower({expression})";
+
+        private static string ToLowerIfNeeded(bool hasTextInfo, RegexOptions options, string expression, bool toLower) => toLower ? ToLower(hasTextInfo, options, expression) : expression;
+
+        private static void EmitTextInfoIfRequired(IndentedTextWriter writer, ref bool textInfoEmitted, ref bool hasTextInfo, RegexMethod rm)
+        {
+            if (textInfoEmitted)
+            {
+                return;
+            }
+            textInfoEmitted = true;
+
+            // Emit local to store current culture if needed
+            if ((((RegexOptions)rm.Options) & RegexOptions.CultureInvariant) == 0)
+            {
+                bool needsCulture = (((RegexOptions)rm.Options) & RegexOptions.IgnoreCase) != 0 || rm.Code.BoyerMoorePrefix?.CaseInsensitive == true;
+                if (!needsCulture && rm.Code.LeadingCharClasses is not null)
+                {
+                    for (int i = 0; i < rm.Code.LeadingCharClasses.Length; i++)
+                    {
+                        if (rm.Code.LeadingCharClasses[i].CaseInsensitive)
+                        {
+                            needsCulture = true;
+                            break;
+                        }
+                    }
+                }
+
+                if (needsCulture)
+                {
+                    hasTextInfo = true;
+                    writer.WriteLine("// IgnoreCase with CultureInfo.CurrentCulture");
+                    writer.WriteLine("TextInfo textInfo = CultureInfo.CurrentCulture.TextInfo;");
+                    writer.WriteLine();
+                }
+            }
+        }
+
+        private static string MatchCharacterClass(bool hasTextInfo, RegexOptions options, string chExpr, string charClass, bool caseInsensitive)
```

Comment:
a method like this is where it's fairly easy to imagine a single implementation that either emits IL or C# as requested.

Reply:
Maybe, maybe not.  Even if it is, this is a tiny part of the whole implementation.

I know you care a lot about such sharing.  If you really want to move that forward, I suggest you invest a dev in prototyping it, and using this implementation as the yardstick to measure its success by.

Reply:
Yup. This code certainly makes that a lot easier: it becomes essentially an exercise in how far we can reasonably reduce code duplication.